build!: Declarative builds via pyproject.toml

gavindsouza opened the pull request:
#### Changes

```diff
+ pyproject.toml
- setup.py
- requirements.txt
```

#### Words

There was an attempt to switch to declarative setup _(2 years ago)_, which was to use
`setup.cfg`. However, we've made more progress in the "world of packaging"
since. Also, given we're using a `pyproject.toml` file already to govern
certain aspects of our project maintenance, it's probably a good time to
bring these things together. bye-bye to clutter and confusion in
maintaining multiple files and formats.

We've used setuptools to package frappe for the longest time. Maybe
since the first editable installation. However, setuptools doesn't support
PEP 660 yet which supports editable installs via pyproject file.

So why now? Primarily because I wanted to start tracking system
dependencies for Frappe. Doing so through pyproject file made the most
sense. This change seemed to be the obvious pre-requisite for the
upcoming changes to our packaging systems.

#### Refs

System dependency tracking: https://github.com/frappe/frappe/issues/13811
Moving app meta to toml: https://github.com/frappe/frappe/issues/15675
Flit docs: https://flit.pypa.io/en/latest/pyproject_toml.html
Previous attempt: https://github.com/frappe/frappe/pull/10773
Which build tool: https://stackoverflow.com/a/69711730/10309266
PEP 660: https://peps.python.org/pep-0660/

gavindsouza committed:
build: Declarative builds via pyproject.toml

There was an attempt to switch to declarative setup, which was to use
setup.cfg. However, we've made more progress in the "world of packaging"
since. Also, given we're using a pyproject file already to govern
certain aspects of our project maintenance, it's probably a good time to
bring these things together. bye-bye to clutter and confusion in
maintaining multiple files and formats.

We've used setuptools to package frappe for the longest time. Maybe
since the first editable installation. Howver setuptools doesn't support
PEP 660 yet which supports editable installs via pyproject file.

So why now? Primarily because I wanted to start tracking system
dependencies for Frappe. Doing so through pyproject file made the most
sense. This change seemed to be the obvious pre-requisite for the
upcoming changes to our packaging systems.

Sys deps tracking: https://github.com/frappe/frappe/issues/13811
Flit docs: https://flit.pypa.io/en/latest/pyproject_toml.html
Previous attempt: https://github.com/frappe/frappe/pull/10773
Which build tool: https://stackoverflow.com/a/69711730/10309266
PEP 660: https://peps.python.org/pep-0660/

gavindsouza commented:
Oops, gotta make bench ready for this first 😿

gavindsouza committed:
fix: Add temporary backwards compatible setup.py

setup.py is maintained so that there is no hard dependency on bench
v5.11.0 and people get enough time to update their CLI. Hope the TODO
adds a subtle reminder for the future ;)

gavindsouza committed:
Merge branch 'develop' into pyproject-filt

gavindsouza commented:
Frappe is now compatible with bench v5.11.0 even without `setup.py` file. Re-added it so that this doesn't affect install for users having older versions of the Bench CLI.

gavindsouza committed:
Merge branch 'develop' into pyproject-filt

ankush commented on pyproject.toml:
```diff
@@ -1,3 +1,97 @@
+requires-python = ">=3.8"
+readme = "README.md"
+dynamic = ["version"]
+dependencies = [
```
Update CI configs to hash pyproject.toml for cache validation

ankush replied:
dev dependencies should ideally also move here 👀

gavindsouza replied:
> dev dependencies should ideally also move here :eyes:

Ideally. Maybe it would look something like https://github.com/frappe/frappe/issues/13811#issuecomment-1155153968. It would go under `[tools.bench.dev-dependencies]` as bench handles this.

gavindsouza replied:
We'll have to defer getting rid of dev-dependencies.txt further until bench v5.12.0 gets widely adopted...and if we're going to miss the v14 release then wait until v15 to do so 🤔

ankush replied:
BTW I doubt many users have their own dev-requirements 😬 (scanned on source graph)

We can fix frappe apps and deprecate this totally too. It was never released (v14 feature)

gavindsouza committed:
ci: Invalidate GHA pip cache on pyproject or setup changes

- Cache invalidated if any of the files change: *requirements.txt,
pyproject.toml, setup.py or setup.cfg
- Updated boilerplate for new apps too

gavindsouza committed:
ci: Remove dependency on requirements.txt for builds

gavindsouza committed:
Merge branch 'develop' into pyproject-filt

gavindsouza commented:
ERPNext CI (just in case...) - https://github.com/frappe/erpnext/actions/runs/2501569276
